core/io/marshalls.cpp

```diff
@@ -869,7 +869,7 @@ Error encode_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len, bo
 
 			double d = p_variant;
 			float f = d;
-			if (double(f) != d) {
```

Comment:
This will cause a double to be encoded unless the lower 29 bits of the double are all zero. How often is this the case anyway?

(I know this is pre-existing code, you've just changed it to use the above check on line 818, so your change is indeed good, but I'm just wondering if this can be improved further)

Reply:
> How often is this the case anyway?

Many Godot functions uses `real_t` (e.g. Vector2/Vector3/almost everything in math/a bunch of stuff in physics).
`real_t` can be defined at compile time as either `float` or `double` (default `float`, can be changed with `REAL_T_IS_DOUBLE` but I never tested it).
So if you encode say `Vector2(1.4,1.3).x` the `x` field will be encoded as float.
eg:
```
prints(var2bytes(.5).size(), 8)
prints(var2bytes(.6).size(), 12)
var v = Vector2(.6, 0)
prints(var2bytes(v.x).size(), 8)
```
The support of this is far from optimal, and there are at least 2 areas where we could do better:
- encoding/decoding Vector directly always encode floats (breaking network transfer when using double `real_t`). We probably want an `encode_real` that enforces the `REAL_T_IS_DOUBLE` directive
- GDScript could support defining floats as `0.6f` (float) beside `0.6` (default to double) which are 2 different values. (@vnen ?)

Reply:
I'm not really fond of adding this `double`/`float` distinction to GDScript. Sounds like an unneeded complication. For me it's simpler to always use `double` (everything is stored in a Variant anyway, it wouldn't save space). AFAIK the plan is to move everything to `double` eventually.

Also, I don't think adding this to GDScript would fix the underlying problem of encoding, since you usually do this to engine types that you can't control from scripting (e.g. if Vector2 uses `double` internally, it doesn't matter if the value in the script is a `float` or not, the type will be coerced).

Reply:
Yeah, you are probably right, optimizing size in networking is very important, but for good optimization one will almost certainly use `StreamPeerBuffer`. Which has `put_float` and `put_double`.